src/layout/workspace.rs

```diff
@@ -426,20 +424,97 @@ impl<W: LayoutElement> Workspace<W> {
         ));
     }
 
+    fn animate_view_offset_to_column(&mut self, current_x: i32, idx: usize) {
+        let new_view_offset = self.compute_new_view_offset_for_column(current_x, idx);
+        self.animate_view_offset(current_x, idx, new_view_offset);
+    }
+
+    fn animate_view_offset_to_column_centered(&mut self, current_x: i32, idx: usize) {
+        if self.columns.is_empty() {
+            self.animate_view_offset_to_column(current_x, idx);
+            return;
+        }
+
+        let col = &self.columns[idx];
+        if col.is_fullscreen {
+            self.animate_view_offset_to_column(current_x, idx);
+            return;
+        }
+
+        let width = col.width();
+
+        // If the column is wider than the working area, then on commit it will be shifted to left
+        // edge alignment by the usual positioning code, so there's no use in doing anything here.
+        if self.working_area.size.w <= width {
+            self.animate_view_offset_to_column(current_x, idx);
+            return;
+        }
+
+        let new_view_offset = -(self.working_area.size.w - width) / 2 - self.working_area.loc.x;
+
+        self.animate_view_offset(current_x, idx, new_view_offset);
+    }
+
     fn activate_column(&mut self, idx: usize) {
         if self.active_column_idx == idx {
             return;
         }
 
         let current_x = self.view_pos();
-        self.animate_view_offset_to_column(current_x, idx);
+        match self.options.center_column_on_move {
+            CenterColumnOnMove::Always => {
+                self.animate_view_offset_to_column_centered(current_x, idx)
+            }
+            CenterColumnOnMove::NoFit => {
+                let width = self.columns[idx].width();
+
+                // Calculate whether the selected column together with the left column fit on the view
+                let fit_left = self
+                    .column_left_idx(idx)
+                    .map(|left_idx| {
+                        let left_width = self.columns[left_idx].width();
+
+                        width + left_width + self.options.gaps < self.view_size.w
```

Comment:
Non-fullscreen columns always operate on the working area (which includes struts and layer-shell). The view size is only for fullscreen columns.